File: crates/klio-cfa/src/analyses/contracts.rs
//! Contract-effect catalogue consumed by the lowering. Spec §12.2.5
//! contracts describe a function's effect on the surrounding flow:
//! a precondition that holds on the post-call path, a lambda that
//! runs a specific number of times, or a smart-cast established by
//! a runtime check.
//!
//! Stdlib contracts live in [`stdlib_contract`] (hardcoded by simple
//! name). User contracts declared via
//! `kotlin.contracts.contract { … }` populate
//! [`USER_INLINE_CONTRACTS`] before lowering — the build pass walks
//! every `inline fun` body once for the contract block and records
//! each `callsInPlace(blockName, EXACTLY_ONCE)` it finds. The
//! lowering then treats a call to that user fn the same way it
//! treats a `let { … }` call: inline the trailing-lambda body into
//! the current block so VIA and smart-cast see the body's effects.

/// One effect a contract imposes on the call site's post-call state.
/// Multiple effects can apply to the same call (e.g. a function
/// that both narrows its first argument and propagates the second
/// argument's refinement).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ContractEffect {
    /// `arg(arg_idx)` is non-null after this call returns normally.
    /// Modeled as an `AssumeNull(eq_null=false)` on the arg's reg.
    AssumeNonNull { arg_idx: usize },
    /// The condition expression at `arg(arg_idx)` holds after the
    /// call returns normally. Any `AssumeIs` / `AssumeNull` /
    /// `AssumeRefEq` refinement the lowering recorded for that
    /// register is replayed on the post-call block.
    AssumePredicate { arg_idx: usize },
}

thread_local! {
    /// User-declared `contract { callsInPlace(p, EXACTLY_ONCE) }`
    /// records, keyed by the inline fn's simple name. Each value
    /// lists the parameter names that are invoked exactly once on
    /// the normal path (Kotlin's `InvocationKind.EXACTLY_ONCE`).
    /// The lowering uses this to extend its trailing-lambda inline
    /// scheme to user contracts so a `val` assigned inside the
    /// lambda is observed as definitely assigned at the call site.
    static USER_INLINE_CONTRACTS: std::cell::RefCell<
        std::collections::HashMap<String, Vec<String>>,
    > = std::cell::RefCell::new(std::collections::HashMap::new());
}

/// Replace the user-contract registry. Called once per module
/// build, before any per-function lowering starts. Passing an
/// empty map effectively clears the registry between modules.
pub fn set_user_inline_contracts(
    map: std::collections::HashMap<String, Vec<String>>,
) {
    USER_INLINE_CONTRACTS.with(|c| *c.borrow_mut() = map);
}

/// Lookup the param names of the user inline fn `name` whose
/// contract declares `callsInPlace(p, EXACTLY_ONCE)`. Empty when no
/// user contract is registered for that name.
#[must_use] 
pub fn user_exactly_once_params(name: &str) -> Vec<String> {
    USER_INLINE_CONTRACTS.with(|c| c.borrow().get(name).cloned().unwrap_or_default())
}

/// Lookup table for stdlib functions that participate in contract
/// effects. The returned slice lists every effect to emit on the
/// post-call path.
#[must_use]
pub fn stdlib_contract(name: &str) -> &'static [ContractEffect] {
    use ContractEffect::{AssumeNonNull, AssumePredicate};
    // The slice values live in `static` storage so callers can hand
    // them to the lowering without lifetime gymnastics.
    static NONNULL: &[ContractEffect] = &[AssumeNonNull { arg_idx: 0 }];
    static REQUIRE: &[ContractEffect] = &[AssumePredicate { arg_idx: 0 }];
    match name {
        "requireNotNull" | "checkNotNull" => NONNULL,
        "require" | "check" => REQUIRE,
        _ => &[],
    }
}

File: src/cfa/analyses/contracts.zig
//! Contract-effect catalogue consumed by the lowering: a precondition holding
//! on the post-call path, a lambda that runs a fixed number of times, or a smart
//! cast from a runtime check. Stdlib contracts live in `stdlibContract`; a user
//! `kotlin.contracts.contract { ... }` populates the registry before lowering,
//! and a call to such a function lowers the way `let { ... }` does.

const std = @import("std");

pub const ContractEffect = union(enum) {
    /// Modeled as an `AssumeNull(eq_null=false)` on the argument's register.
    AssumeNonNull: struct { arg_idx: usize },
    /// The condition at `arg(arg_idx)` holds after a normal return: any
    /// refinement recorded for that register is replayed on the post-call block.
    AssumePredicate: struct { arg_idx: usize },

    pub fn eql(self: ContractEffect, other: ContractEffect) bool {
        if (@as(std.meta.Tag(ContractEffect), self) != @as(std.meta.Tag(ContractEffect), other)) {
            return false;
        }
        return switch (self) {
            .AssumeNonNull => |e| e.arg_idx == other.AssumeNonNull.arg_idx,
            .AssumePredicate => |e| e.arg_idx == other.AssumePredicate.arg_idx,
        };
    }
};

/// User `contract { callsInPlace(p, EXACTLY_ONCE) }` records, keyed by the
/// inline function's simple name, each value listing the parameters invoked
/// exactly once on the normal path. Lowering extends its trailing-lambda
/// inlining to them, so a `val` assigned inside the lambda is definitely
/// assigned at the call site.
pub const UserInlineContracts = std.StringHashMap([]const []const u8);

/// Module-level state under a single-build-at-a-time contract: the build driver
/// installs the registry before lowering starts.
var user_inline_contracts: ?UserInlineContracts = null;

/// Once per module build, before any per-function lowering; an empty map clears
/// it. Takes ownership of `map` and frees any prior registry.
pub fn setUserInlineContracts(map: UserInlineContracts) void {
    if (user_inline_contracts) |*old| old.deinit();
    user_inline_contracts = map;
}

/// Empty when no user contract is registered for `name`.
pub fn userExactlyOnceParams(name: []const u8) []const []const u8 {
    if (user_inline_contracts) |*c| {
        if (c.get(name)) |params| return params;
    }
    return &.{};
}

pub fn resetForTest() void {
    if (user_inline_contracts) |*m| {
        m.deinit();
        user_inline_contracts = null;
    }
}

/// Every effect to emit on the post-call path of a stdlib function.
pub fn stdlibContract(name: []const u8) []const ContractEffect {
    const nonnull = &[_]ContractEffect{.{ .AssumeNonNull = .{ .arg_idx = 0 } }};
    const require = &[_]ContractEffect{.{ .AssumePredicate = .{ .arg_idx = 0 } }};
    if (std.mem.eql(u8, name, "requireNotNull") or std.mem.eql(u8, name, "checkNotNull")) {
        return nonnull;
    }
    if (std.mem.eql(u8, name, "require") or std.mem.eql(u8, name, "check")) {
        return require;
    }
    return &.{};
}

const testing = std.testing;

test {
    testing.refAllDecls(@This());
}

test "stdlib contract lookup" {
    const rn = stdlibContract("requireNotNull");
    try testing.expectEqual(@as(usize, 1), rn.len);
    try testing.expect(rn[0] == .AssumeNonNull);
    try testing.expectEqual(@as(usize, 0), rn[0].AssumeNonNull.arg_idx);

    const cnn = stdlibContract("checkNotNull");
    try testing.expect(cnn[0] == .AssumeNonNull);

    const req = stdlibContract("require");
    try testing.expect(req[0] == .AssumePredicate);
    const chk = stdlibContract("check");
    try testing.expect(chk[0] == .AssumePredicate);

    try testing.expectEqual(@as(usize, 0), stdlibContract("unknown").len);
}

test "user inline contracts round-trip" {
    defer resetForTest();
    try testing.expectEqual(@as(usize, 0), userExactlyOnceParams("run").len);

    var map = UserInlineContracts.init(testing.allocator);
    const params = [_][]const u8{"block"};
    try map.put("withResource", &params);
    setUserInlineContracts(map);

    const got = userExactlyOnceParams("withResource");
    try testing.expectEqual(@as(usize, 1), got.len);
    try testing.expectEqualStrings("block", got[0]);
    try testing.expectEqual(@as(usize, 0), userExactlyOnceParams("missing").len);

    setUserInlineContracts(UserInlineContracts.init(testing.allocator));
    try testing.expectEqual(@as(usize, 0), userExactlyOnceParams("withResource").len);
}

test "contract effect equality" {
    const a: ContractEffect = .{ .AssumeNonNull = .{ .arg_idx = 0 } };
    const b: ContractEffect = .{ .AssumeNonNull = .{ .arg_idx = 0 } };
    const c: ContractEffect = .{ .AssumeNonNull = .{ .arg_idx = 1 } };
    const d: ContractEffect = .{ .AssumePredicate = .{ .arg_idx = 0 } };
    try testing.expect(a.eql(b));
    try testing.expect(!a.eql(c));
    try testing.expect(!a.eql(d));
}
